### PR TITLE
feat: allow to pass all options from `chokidar`

### PR DESCRIPTION
**Pull request by snitin315**
- [ ] This is a **bugfix**
- [x] This is a **feature**
- [ ] This is a **code refactor**
- [x] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?
Updated snapshots.

### Motivation / Use-Case

 - Better defaults 
 - allow to pass all options from `chokidar` for `static.watch` and `watchFiles.options`
 

### Breaking Changes
None

### Additional Info
No

### CODE REVIEW

File: lib/Server.js [hunk 1]
```diff
@@ -2143,17 +2167,16 @@ class Server {
...
-      ignoreInitial: true,
-      persistent: true,
-      followSymlinks: false,
-      atomic: false,
-      alwaysStat: true,
-      ignorePermissionErrors: true,
+      ...watchOptions,
       ignored: watchOptions.ignored,
       usePolling,
       interval,
```
Comment (alexander-akait) [inline lib/Server.js:2173]: Please add tests there we override `usePolling` and `interval`

File: lib/Server.js [hunk 2]
```diff
@@ -445,13 +445,22 @@ class Server {
...
     const watchOptions = compilerOptions.watchOptions || {};
+    const defaultWatchOptions = {
+      ignoreInitial: true,
+      persistent: true,
+      followSymlinks: false,
+      atomic: false,
+      alwaysStat: true,
+      ignorePermissionErrors: true,
+      ...watchOptions,
+    };
```
Comment (alexander-akait) [inline lib/Server.js:456]: Let's move normalization from https://github.com/webpack/webpack-dev-server/blob/b4a0cfd326cfa73a4c932254244df44a3deec1a1/lib/Server.js#L2156 to here, wecan calculate default `usePolling` and `interval` value here

File: lib/Server.js [hunk 3]
```diff
@@ -444,16 +444,49 @@ class Server {
...
+        ...compilerWatchOptions,
+        ...watchOptions,
+        ignored: watchOptions.ignored,
+        usePolling,
+        interval,
+      };
+
+      if (finalWatchOptions.poll) {
+        delete finalWatchOptions.poll;
+      }
```
Comment (alexander-akait) [inline lib/Server.js:483]: Avoid it, you already know about `poll` is not valid for chokidar, only pass valid options

Comment (snitin315) [reply lib/Server.js:483]: `poll` can also come from `compilerWatchOptions`.

Comment (alexander-akait) [reply lib/Server.js:483]: Just not put `poll` in your default options

File: lib/Server.js [hunk 4]
```diff
@@ -464,24 +468,23 @@ class Server {
           : // eslint-disable-next-line no-undefined
             undefined;
 
+      if (watchOptions.poll) {
+        delete watchOptions.poll;
+      }
```
Comment (alexander-akait) [inline lib/Server.js:473]: Why not get only valid options from `watchOptions`?

Comment (snitin315) [reply lib/Server.js:473]: Because we respect `poll` option for normalizing `usePolling` and `interval` option.

Comment (alexander-akait) [reply lib/Server.js:473]: Using this will modify webpack compiler options, you should not touch them

Comment (alexander-akait) [reply lib/Server.js:473]: You need `Default Options` + `Options from watchOptions` + `DeveloperOptions`, so developer can override even `usePolling` and `interval` (but by default we get them from `watchOptions`) and other

Comment (snitin315) [reply lib/Server.js:473]: Here we are not modifying `compiler.options.watchOptions`, only `devServer.static.watch` and `devServer.watchFiles.options`

Comment (snitin315) [reply lib/Server.js:473]: got it.

Comment (snitin315) [reply lib/Server.js:473]: Previously we were also respecting `devServer.static.watch.poll` which is not valid.

Comment (alexander-akait) [reply lib/Server.js:473]: We should respect them too, i.e.:
```
{
  ...defaultOptions,
  ...watchOptions
  ...developerOptionsFroChokidar
}
```

Comment (snitin315) [reply lib/Server.js:473]: This will pass `poll` to `chokidar` when developer defines `devServer.static.watch.poll`.

**Pull request merged**